pkg/broker/ingress/handler.go

```diff
@@ -18,97 +18,131 @@ package ingress
 type DecoupleSink interface {
-	// Send sends the event to a decoupling sink.
-	Send(ctx context.Context, event cev2.Event) protocol.Result
+	// Send sends the event from a broker to the corresponding decoupling sink.
+	Send(ctx context.Context, ns, broker string, event cev2.Event) protocol.Result
 }
 
 // handler receives events and persists them to storage (pubsub).
+// TODO(liu-cong) add metrics
+// TODO(liu-cong) add tracing
+// TODO(liu-cong) support event TTL
 type handler struct {
-	// inbound is the cloudevents client to receive events.
-	inbound cev2.Client
+	// httpReceiver starts an HTTP server to receive events.
+	httpReceiver *kncloudevents.HttpMessageReceiver
 	// decouple is the client to send events to a decouple sink.
 	decouple DecoupleSink
+	logger   *zap.Logger
 }
 
-func NewHandler(ctx context.Context, options ...Option) (*handler, error) {
-	h := &handler{}
+// NewHandler creates a new ingress handler.
+func NewHandler(ctx context.Context, options ...HandlerOption) (*handler, error) {
+	h := &handler{
+		logger: logging.FromContext(ctx),
+	}
+
 	for _, option := range options {
 		option(h)
 	}
 
-	var err error
-	if h.inbound == nil {
-		if h.inbound, err = newDefaultHTTPClient(); err != nil {
-			return nil, fmt.Errorf("failed to create inbound cloudevent client: %w", err)
-		}
+	if h.httpReceiver == nil {
+		h.httpReceiver = kncloudevents.NewHttpMessageReceiver(defaultPort)
 	}
 
 	if h.decouple == nil {
-		if h.decouple, err = h.newDefaultPubSubClient(ctx); err != nil {
-			return nil, fmt.Errorf("failed to create decouple cloudevent client: %w", err)
+		sink, err := NewMultiTopicDecoupleSink(ctx)
+		if err != nil {
+			return nil, errors.Wrap(err, "creating default multi topic decouple sink")
 		}
+		h.decouple = sink
 	}
 
 	return h, nil
 }
 
+// Start blocks to receive events over HTTP.
 func (h *handler) Start(ctx context.Context) error {
-	return h.inbound.StartReceiver(ctx, h.receive)
+	return h.httpReceiver.StartListen(ctx, h)
 }
 
-// receive receives events from inbound and sends them to decouple.
-func (h *handler) receive(ctx context.Context, event cev2.Event) protocol.Result {
-	// TODO(liu-cong) add metrics
-	// TODO(liu-cong) route events based on broker.
+// ServeHTTP implements net/http Handler interface method.
+// 1. Performs basic validation of the request.
+// 2. Parse request URL to get namespace and broker.
+// 3. Convert request to event.
+// 4. Send event to decouple sink.
+func (h *handler) ServeHTTP(response nethttp.ResponseWriter, request *nethttp.Request) {
+	h.logger.Debug("Serving http", zap.Any("request", request))
+	if request.Method != nethttp.MethodPost {
+		response.WriteHeader(nethttp.StatusMethodNotAllowed)
+		return
+	}
 
-	if res := h.decouple.Send(ctx, event); !cev2.IsACK(res) {
-		logging.FromContext(ctx).Error("Error publishing to PubSub", zap.String("event", event.String()), zap.Error(res))
-		return res
+	// Path should be in the form of "/<ns>/<broker>"
```

Comment:
```suggestion
	// Path should be in the form of "/<ns>/<broker>".
```

I also recommend including the Broker's UID if you can. It just makes it less likely someone sends to the wrong Broker by mistake.

Reply:
@grantr Any concern from the controller perspective?

Reply:
I have concerns from the UX perspective. If we use a UID in the path then:
- The path to produce to the Broker is no longer predictable
- There is no way to configure the producer with a broker URI in advance of the Broker being created
- If the Broker is deleted and recreated, the producer must update its broker URI

None of these are issues for producers that have a watch on the Broker and can resolve its address, but watching the Broker requires permissions, it requires a control loop, it requires a lot of lifecycle code just to get that URI. I'd like to keep open the option to start a random pod and send to a known URI.

If the concern about sending to the wrong broker by mistake is about the "delete and recreate the Broker" case, then I'd argue that it makes sense for the new Broker to have the same URI because the producer's directive is still "send an event to the Broker with this name".

if the concern is about sending to the wrong broker due to a misconfiguration, then I don't think including the UID helps, because the user specifies the Broker by name anyway in the reference (for CRD sources).

So my inclination is to not use the URI in the path.

Reply:
I retract my suggestion. The predictability of the Broker's URI was a conscious design choice of the Broker. I did not think through my suggestion, which would have broken that design.